ros_packages/voice_assistant/voice_assistant/token_service.py

```diff
@@ -0,0 +1,168 @@
+from datatypes.srv import EncryptToken, DecryptToken, ExistToken
+from rclpy.executors import SingleThreadedExecutor
+from rclpy.node import Node
+from std_msgs.msg import String, Empty
+
+from . import SECRETS_DIR, SALT_PATH, TOKEN_PATH
+
+
+class TokenServiceNode(Node):
+    def __init__(self):
+        super().__init__("token_service")
+        self.active_token: Union[str, None] = None
+        self.is_token_stored: bool = self._check_if_previous_token_stored()
+
+        self.delete_token_subscription = self.create_subscription(
+            Empty, "delete_token", self.delete_token_callback, 10
+        )
+        self.token_exists_service = self.create_service(
+            ExistToken, "exist_token", self.token_exists_callback
+        )
+        self.encryption_service = self.create_service(
+            EncryptToken, "encrypt_token", self.encrypt_token_callback
+        )
+        self.decryption_service = self.create_service(
+            DecryptToken, "decrypt_token", self.decrypt_token_callback
+        )
+        self.token_publisher = self.create_publisher(String, "public_api_token", 10)
+
+        self.get_logger().info("Now running TOKEN SERVICE")
+
+    def publish_token(self, token: str) -> None:
+        self.active_token = token
+        msg = String()
+        msg.data = token
+        self.token_publisher.publish(msg)
+
+    def delete_token_callback(self, _: Empty) -> None:
+        try:
+            if self._check_if_previous_token_stored():
+                os.remove(TOKEN_PATH)
+                os.remove(SALT_PATH)
+        except Exception as e:
+            self.get_logger().warn(f"Token file could not be deleted: {e}")
+
+        # even if token stored in file cannot be deleted, remove the active token
+        self.is_token_stored = False
+        self.publish_token("")
+
+    def token_exists_callback(
+        self, _: ExistToken.Request, response: ExistToken.Response
+    ):
+        response.token_exists = self.is_token_stored
+        response.token_active = False
+        if self.active_token:
+            response.token_active = True
+        return response
+
+    def encrypt_token_callback(
+        self, request: EncryptToken.Request, response: EncryptToken.Response
+    ):
+        token: str = request.token
+        password: str = request.password
+
+        is_successful: bool = self.encrypt_token(token, password)
+        if is_successful:
+            self.publish_token(token)
+
+        self.is_token_stored = is_successful
+        response.is_successful = is_successful
+        return response
+
+    def decrypt_token_callback(
+        self, request: DecryptToken.Request, response: DecryptToken.Response
+    ):
+        password: str = request.password
+        is_successful: bool = True
+
+        try:
+            token: str = self.decrypt_token(password)
+            self.publish_token(token)
+        except Exception:
+            self.get_logger().warn(f"Token could not be decrypted")
```

Comment:
`error` normally indicates a fatal error in the operation, but failing to decrypt a token is entirely possible and expected if the user inputs a wrong password. That's why I think `warn` fits better here, as it might be a real exception, but in most cases won't be.